testMaxSizeExceededOnResize fails to create array to resize

There have been a few failures recently of `BigArraysTests#testMaxSizeExceededOnResize` in which we hit a `CircuitBreakingException` on this line:

https://github.com/elastic/elasticsearch/blob/46ac8d1a51339972f9647775155dc1a2ee1ca624/server/src/test/java/org/elasticsearch/common/util/BigArraysTests.java#L363

The full stack trace from https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+master+multijob-unix-compatibility/os=debian/1/console follows:

```
java.lang.reflect.InvocationTargetException
	at __randomizedtesting.SeedInfo.seed([9ECB54FC37009C71:A6FDADAB65815066]:0)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at org.elasticsearch.common.util.BigArraysTests.testMaxSizeExceededOnResize(BigArraysTests.java:363)
	at sun.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at sun.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at sun.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.lang.reflect.Method.invoke(Method.java:498)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.invoke(RandomizedRunner.java:1750)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$8.evaluate(RandomizedRunner.java:938)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$9.evaluate(RandomizedRunner.java:974)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$10.evaluate(RandomizedRunner.java:988)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleSetupTeardownChained$1.evaluate(TestRuleSetupTeardownChained.java:49)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at org.apache.lucene.util.TestRuleThreadAndTestName$1.evaluate(TestRuleThreadAndTestName.java:48)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl.forkTimeoutingTask(ThreadLeakControl.java:817)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$3.evaluate(ThreadLeakControl.java:468)
	at com.carrotsearch.randomizedtesting.RandomizedRunner.runSingleTest(RandomizedRunner.java:947)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$5.evaluate(RandomizedRunner.java:832)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$6.evaluate(RandomizedRunner.java:883)
	at com.carrotsearch.randomizedtesting.RandomizedRunner$7.evaluate(RandomizedRunner.java:894)
	at org.apache.lucene.util.AbstractBeforeAfterRule$1.evaluate(AbstractBeforeAfterRule.java:45)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleStoreClassName$1.evaluate(TestRuleStoreClassName.java:41)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.NoShadowingOrOverridesOnMethodsRule$1.evaluate(NoShadowingOrOverridesOnMethodsRule.java:40)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at org.apache.lucene.util.TestRuleAssertionsRequired$1.evaluate(TestRuleAssertionsRequired.java:53)
	at org.apache.lucene.util.TestRuleMarkFailure$1.evaluate(TestRuleMarkFailure.java:47)
	at org.apache.lucene.util.TestRuleIgnoreAfterMaxFailures$1.evaluate(TestRuleIgnoreAfterMaxFailures.java:64)
	at org.apache.lucene.util.TestRuleIgnoreTestSuites$1.evaluate(TestRuleIgnoreTestSuites.java:54)
	at com.carrotsearch.randomizedtesting.rules.StatementAdapter.evaluate(StatementAdapter.java:36)
	at com.carrotsearch.randomizedtesting.ThreadLeakControl$StatementRunner.run(ThreadLeakControl.java:368)
	at java.lang.Thread.run(Thread.java:748)
Caused by: CircuitBreakingException[[parent] Data too large, data for [<reused_arrays>] would be [491472096/468.7mb], which is larger than the limit of [489108275/466.4mb], real usage: [491390176/468.6mb], new bytes reserved: [81920/80kb]]
	at org.elasticsearch.indices.breaker.HierarchyCircuitBreakerService.checkParentLimit(HierarchyCircuitBreakerService.java:300)
	at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.addEstimateBytesAndMaybeBreak(ChildMemoryCircuitBreaker.java:128)
	at org.elasticsearch.common.util.BigArrays.adjustBreaker(BigArrays.java:407)
	at org.elasticsearch.common.util.BigArrays.newByteArray(BigArrays.java:474)
	at org.elasticsearch.common.util.BigArrays.newByteArray(BigArrays.java:489)
	... 42 more
```

This may relate to 46ac8d1a51339972f9647775155dc1a2ee1ca624, although there seem to have been failures of this nature before that was merged. Unfortunately the build logs for the similar-looking failures were lost in the move so it's not clear.

The `REPRODUCE WITH` line follows, although it does not reproduce for me.

```
REPRODUCE WITH: ./gradlew :server:test \
  -Dtests.seed=9ECB54FC37009C71 \
  -Dtests.class=org.elasticsearch.common.util.BigArraysTests \
  -Dtests.method="testMaxSizeExceededOnResize" \
  -Dtests.security.manager=true \
  -Dtests.locale=cs \
  -Dtests.timezone=Australia/LHI \
  -Dcompiler.java=10 \
  -Druntime.java=8
```

@jpountz could you have a look at this?

I just ran into this again in a PR on a branch that contains  b06a082: https://elasticsearch-ci.elastic.co/job/elastic+elasticsearch+pull-request/439/console

```
18:13:35    > Throwable #1: java.lang.reflect.InvocationTargetException
18:13:35    > 	at __randomizedtesting.SeedInfo.seed([A5BCCF97E347CCDA:9D8A36C0B1C600CD]:0)
18:13:35    > 	at org.elasticsearch.common.util.BigArraysTests.testMaxSizeExceededOnResize(BigArraysTests.java:364)
18:13:35    > 	at java.lang.Thread.run(Thread.java:748)
18:13:35    > Caused by: CircuitBreakingException[[request] Data too large, data for [<reused_arrays>] would be [2818048/2.6mb], which is larger than the limit of [2817079/2.6mb]]
18:13:35    > 	at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.circuitBreak(ChildMemoryCircuitBreaker.java:98)
18:13:35    > 	at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.limit(ChildMemoryCircuitBreaker.java:170)
18:13:35    > 	at org.elasticsearch.common.breaker.ChildMemoryCircuitBreaker.addEstimateBytesAndMaybeBreak(ChildMemoryCircuitBreaker.java:123)
18:13:35    > 	at org.elasticsearch.common.util.BigArrays.adjustBreaker(BigArrays.java:407)
18:13:35    > 	at org.elasticsearch.common.util.BigArrays.newLongArray(BigArrays.java:608)
18:13:35    > 	at org.elasticsearch.common.util.BigArrays.newLongArray(BigArrays.java:623)
18:13:35    > 	... 42 more
```

Looks very similar, so I'm reopening for further investigation.

--- a/server/src/test/java/org/elasticsearch/common/util/BigArraysTests.java
+++ b/server/src/test/java/org/elasticsearch/common/util/BigArraysTests.java
@@ -351,7 +351,7 @@ public class BigArraysTests extends ESTestCase {
 
     public void testMaxSizeExceededOnResize() throws Exception {
         for (String type : Arrays.asList("Byte", "Int", "Long", "Float", "Double", "Object")) {
-            final int maxSize = randomIntBetween(1 << 10, 1 << 22);
+            final int maxSize = randomIntBetween(1 << 8, 1 << 14);
             HierarchyCircuitBreakerService hcbs = new HierarchyCircuitBreakerService(
                     Settings.builder()
                             .put(REQUEST_CIRCUIT_BREAKER_LIMIT_SETTING.getKey(), maxSize, ByteSizeUnit.BYTES)
@@ -360,12 +360,18 @@ public class BigArraysTests extends ESTestCase {
                     new ClusterSettings(Settings.EMPTY, ClusterSettings.BUILT_IN_CLUSTER_SETTINGS));
             BigArrays bigArrays = new BigArrays(null, hcbs, false).withCircuitBreaking();
             Method create = BigArrays.class.getMethod("new" + type + "Array", long.class);
-            final int size = scaledRandomIntBetween(10, maxSize / 8);
+            final int size = scaledRandomIntBetween(10, maxSize / 16);
             BigArray array = (BigArray) create.invoke(bigArrays, size);
             Method resize = BigArrays.class.getMethod("resize", array.getClass().getInterfaces()[0], long.class);
-            final long newSize = maxSize + 1;
-            InvocationTargetException e = expectThrows(InvocationTargetException.class, () -> resize.invoke(bigArrays, array, newSize));
-            assertTrue(e.getCause() instanceof CircuitBreakingException);
+            while (true) {
+                long newSize = array.size() * 2;
+                try {
+                    array = (BigArray) resize.invoke(bigArrays, array, newSize);
+                } catch (InvocationTargetException e) {
+                    assertTrue(e.getCause() instanceof CircuitBreakingException);
+                    break;
+                }
+            }
             assertEquals(array.ramBytesUsed(), hcbs.getBreaker(CircuitBreaker.REQUEST).getUsed());
             array.close();
             assertEquals(0, hcbs.getBreaker(CircuitBreaker.REQUEST).getUsed());

